Implement the list service bindings endpoint

acosta11 opened the pull request:
## Is there a related GitHub Issue?
#603

## What is this change about?
#603 

## Does this PR introduce a breaking change?
No

## Acceptance Steps
#603

## Tag your pair, your PM, and/or team
paired w/ @julian-hj @gnovv

julian-hj reviewed (approved):
🦸

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+			_ = createServiceInstance(context.Background(), k8sClient, serviceInstanceGUID, spaceGUID, "service-instance-name", secretName)
+		})
+
+		When("type is `app`", func() {
```
What other types are there? If `app` is the only option, this `When` is misleading and not needed

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+		When("type is `app`", func() {
+			It("is successful and creates a CFServiceBinding resource", func() {
+				var err error
+				req, err = http.NewRequestWithContext(ctx, "POST", serverURI("/v3/service_credential_bindings"), strings.NewReader(fmt.Sprintf(`{
```
Invoking the method under test should generally go into `JustBeforeEach` instead of being inlined in an `It`

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+					"details":          HaveKeyWithValue("href", HaveSuffix("/v3/service_credential_bindings/"+guid+"/details")),
+				})))
+
+				serviceBinding := new(servicesv1alpha1.CFServiceBinding)
```
The test so far has been checking the response. From here on the test checks the k8s object. Therefore, that should be a separate `It`. Also, if creating the object is performed in a `JustBeforeEach`, that It would be just a sibling to the response checking one.

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+		)
+
+		BeforeEach(func() {
+			org := createOrgAnchorAndNamespace(ctx, rootNamespace, generateGUID())
```
Creating the org, the space and the app is shared for both `Describes`. In. order to remove the code duplication the common setup should be extracted into a top-level `BeforeEach`

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+			_ = createServiceBinding(ctx, k8sClient, serviceBinding2GUID, spaceGUID, nil, serviceInstance2.Name, cfApp.Name)
+
+			var err error
+			req, err = http.NewRequestWithContext(ctx, "GET", serverURI("/v3/service_credential_bindings"), strings.NewReader(""))
```
The request is not changed in the underlying contexts, so maybe moving its creation to the `JustBeforeEach` would maybe make the code a bit more readable.

danail-branekov commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,280 @@
+			router.ServeHTTP(rr, req)
+		})
+
+		When("the user has permission to List ServiceBindings", func() {
```
There is no test for the case when the user does not have permissions to list service bindings.

danail-branekov commented on api/apis/service_binding_handler.go:
```diff
@@ -93,6 +96,49 @@ func (h *ServiceBindingHandler) createHandler(authInfo authorization.Info, w htt
+	listFilter := new(payloads.ServiceBindingList)
+	err := schema.NewDecoder().Decode(listFilter, r.Form)
+	if err != nil {
+		switch err.(type) {
```
How about extracting the check for UnknownKeyError into a utility function so that the code below reads like

```
if isUnknownKeyError(err) {
   // return unknown key error
}
// return unknown error

acosta11 replied:
Added https://github.com/cloudfoundry/cf-k8s-controllers/issues/698

danail-branekov commented on api/apis/service_binding_handler_test.go:
```diff
@@ -365,4 +366,124 @@ var _ = Describe("ServiceBindingHandler", func() {
 	})
+
+	Describe("the GET /v3/service_credential_bindings endpoint", func() {
+		makeListRequest := func(queryParams ...string) {
```
How about renaming `makeListRequest` to `JustBeforeEach` so that the test is more idiomatic. then the function itself would not have to be explicitly invoked in all `BeforeEach`s

danail-branekov commented on api/repositories/service_binding_repository.go:
```diff
@@ -142,3 +152,57 @@ func cfServiceBindingToRecord(binding v1alpha1.CFServiceBinding) ServiceBindingR
+func (r *ServiceBindingRepo) ListServiceBindings(ctx context.Context, authInfo authorization.Info, message ListServiceBindingsMessage) ([]ServiceBindingRecord, error) {
+	nsList, err := r.namespacePermissions.GetAuthorizedSpaceNamespaces(ctx, authInfo)
+	if err != nil {
+		// untested
```
Is there any reason to have this branch untested? Why do we need such sort of comments?

danail-branekov commented on api/repositories/service_binding_repository.go:
```diff
@@ -142,3 +152,57 @@ func cfServiceBindingToRecord(binding v1alpha1.CFServiceBinding) ServiceBindingR
+		filteredServiceBindings = append(filteredServiceBindings, applyServiceBindingListFilter(serviceInstanceList.Items, message)...)
+	}
+
+	return returnServiceBindingList(filteredServiceBindings), nil
```
`return returnXXX` sounds weird, maybe change the name to describe what the function actually does, e.g. `toServiceBindingRecords`

danail-branekov commented on api/repositories/service_binding_repository_test.go:
```diff
@@ -237,7 +243,103 @@ var _ = Describe("ServiceBindingRepo", func() {
+				createRoleBinding(testCtx, userName, spaceDeveloperClusterRole.Name, space2.Name)
+			})
+
+			Describe("no query parameters are specified", func() {
```
In order to be consistent with the context below, this `Describe` should be a `When`

danail-branekov commented on api/repositories/service_binding_repository_test.go:
```diff
@@ -237,7 +243,103 @@ var _ = Describe("ServiceBindingRepo", func() {
+
+			Describe("no query parameters are specified", func() {
+				It("returns a list of ServiceBindingRecords in the spaces the user has access to", func() {
+					Eventually(func() []repositories.ServiceBindingRecord {
```
Does this need to be `Eventually` instead of ordinary `Expect`? Resources are created synchronously in the `BeforeEach`, so should be immediately available in the `It`

acosta11 replied:
I don't know that the resources are actually synchronously available, even after the create call succeeds,  though perhaps the cache gets updated if the element is absent? I feel like we've seen flakes related to directly calling `Expect`.

danail-branekov commented on api/repositories/service_binding_repository_test.go:
```diff
@@ -237,7 +243,103 @@ var _ = Describe("ServiceBindingRepo", func() {
+
+				It("eventually returns only a subset of the ServiceBindings that match the provided filter", func() {
+					Eventually(func() []repositories.ServiceBindingRecord {
+						responseServiceBindings, err := repo.ListServiceBindings(context.Background(), authInfo, requestMessage)
```
`requestMessage` is a local to this invocation, so could be inlined, instead of initialising it in a `BeforeEach` above. Alternatively, the `ListServiceBindings` function could be invoked in `JustBeforeEach` so that it is not invoked in `Its`. Then, setting up the request message in `BeforeEach` above would make sense.

danail-branekov commented on api/repositories/service_binding_repository_test.go:
```diff
@@ -237,7 +243,103 @@ var _ = Describe("ServiceBindingRepo", func() {
+
+		Describe("the user does not have access to any namespaces", func() {
+			It("always returns an empty list and no error", func() {
+				Consistently(func() []repositories.ServiceBindingRecord {
```
Same as the above comment regarding `Expect` vs `Eventually`

danail-branekov commented on api/repositories/service_instance_repository_test.go:
```diff
@@ -504,6 +504,7 @@ func initializeServiceInstanceCreateMessage(serviceInstanceName string, spaceGUI
 	}
 }
 
+// TODO: refactor to use createServiceInstanceCR
```
Is there a a chore about this TODO? If yes, having it here does not make sense, if no, please create one.

acosta11 replied:
Good call, we just did the refactor in place since the changes were only in one file.

danail-branekov reviewed (changes requested):
Please look at the comments above.

acosta11 commented:
Put the PR back in draft to prevent merges, but addressed all of the comments. Need to squash and do one last look-over, but we have now hit EOD and some.

georgethebeatle commented on api/apis/integration/service_binding_test.go:
```diff
@@ -0,0 +1,295 @@
+			serviceInstanceGUID = generateGUID()
+			_ = createServiceInstance(context.Background(), k8sClient, serviceInstanceGUID, spaceGUID, "service-instance-name", secretName)
+			req, err = http.NewRequestWithContext(ctx, "POST", serverURI("/v3/service_credential_bindings"), strings.NewReader(fmt.Sprintf(`{
+																																		 "relationships": {
```
This formatting is a bit weird. On our machine it is offset in the far right part of the screen and hard to read.

georgethebeatle commented:
Overall looks good. We filed a draft PR #708 that suggests some further improvements. Please have a look.
